server/server.py

```diff
@@ -48,16 +55,93 @@ def exec_DB(sql, params=()):
     conn.close()
 
 
+def get_admin_password():
+    result = query_DB("SELECT password FROM users WHERE name='admin'")
+    if result:
+        return result[0][0]
+    else:
+        return None
+
+
+def set_admin_password(admin_password):
+    password_hash = hashlib.sha256()
+    password_hash.update(admin_password)
+    exec_DB("DELETE FROM users WHERE name='admin'")
+    exec_DB("INSERT INTO users VALUES (?, ?, ?)", (None, "admin", password_hash.hexdigest()))
+
+
+def require_admin(func):
+        def wrapper(*args, **kwargs):
+            global session_cookie
+            global last_session_activity
+            global SESSION_TIMEOUT
+            if session_cookie and COOKIE_NAME in cherrypy.request.cookie and session_cookie == cherrypy.request.cookie[COOKIE_NAME].value:
+                if time.time() - last_session_activity > SESSION_TIMEOUT:
+                    return app.disconnect()
+                else:
+                    last_session_activity = time.time()
+                    return func(*args, **kwargs)
+            else:
+                raise cherrypy.HTTPRedirect("/login")
+        return wrapper
+
+
 class Main(object):
     @cherrypy.expose
+    @require_admin
     def index(self):
         with open("Menu.html", "r") as f:
             html = f.read()
             return html
 
+    @cherrypy.expose
+    def login(self, password=''):
+        admin_password = get_admin_password()
+        if not admin_password:
+            if password:
+                set_admin_password(password)
+                return 'Admin password set successfully. <a href="./login">Click here to login</a>'
+            else:
+                with open("CreatePassword.html", "r") as f:
+                    html = f.read()
+                    return html
+        else:
+            password_hash = hashlib.sha256()
+            password_hash.update(password)
+            if password_hash.hexdigest() == get_admin_password():
+                global session_cookie
+                session_cookie = ''.join(random.choice(string.ascii_letters + string.digits) for i in range(64))
+                cherrypy.response.cookie[COOKIE_NAME] = session_cookie
+                global last_session_activity
+                last_session_activity = time.time()
+                raise cherrypy.HTTPRedirect("/")
+            else:
+                with open("Login.html", "r") as f:
+                    html = f.read()
+                    return html
+
+    @cherrypy.expose
+    def disconnect(self):
+        session_cookie = None
+        cherrypy.response.cookie[COOKIE_NAME] = ''
+        cherrypy.response.cookie[COOKIE_NAME]['expires'] = 0
+        return 'You have been disconnected. <a href="./login">Click here to login</a>'
+
+    @cherrypy.expose
+    @require_admin
+    def passchange(self, password=''):
```

Comment:
this method seems to work but it's hidden. You should add a link to it in the menu like for `disconnect`.